### PR TITLE
Host helper C++ class

### PR DESCRIPTION
**Pull request by Trinitou**
Adding a C++ helper class CLAP host developers can easily derive from

Left out here for future PRs / discussion
- misbehaviour checks, e.g. some ensureMainThread / ensureAudioThread methods, checking level templating etc.
- clap-helpers repository description: "Minimal set of helpers to implement your plugin" would need to change to something more welcoming also for host devs -> or should this be split into two separate repos instead ("clap-plugin-helpers" and "clap-host-helpers")?
- documentation
- test

### CODE REVIEW

**Comment by Trinitou** [comment]
Probably I'll focus on the `clap::helpers::Host` first and leave out the PluginProxy stuff (can go into a separate PR). Main goal would be to extract the reusable stuff from `PluginHost` in https://github.com/free-audio/clap-host/blob/main/host/plugin-host.hh so that `PluginHost` can be derived from the new `clap::helpers::Host` class

**Comment by abique** [comment]
Good luck with this task, this isn't an easy one. The host'll be harder to generalize/glue than the plugin.

**Comment by Trinitou** [comment]
- [x] reverted the PluginProxy changes
- [x] removed the template code (at least for now)
- [x] added all extensions necessary to derive the PluginHost class from the `free-audio/plugin-host` repo

(Edit: moved open todos into first comment)

Please also have a look into this commit in my fork of the `plugin-host` repo:
[f22ed540aa206c2452bdfb9011a76cf3b14cd9f2](https://github.com/Trinitou/clap-host/commit/f22ed540aa206c2452bdfb9011a76cf3b14cd9f2)
This is how I think it could look like using the new `clap::helpers::Host` class. Maybe some details are still missing. What do you think @abique ?

**Comment by Trinitou** [comment]
I think we could leave it like this for now except you want to have misbehavior checks from the beginning.

**Comment by abique** [comment]
> Please also have a look into this commit in my fork of the `plugin-host` repo:
> [f22ed540aa206c2452bdfb9011a76cf3b14cd9f2](https://github.com/Trinitou/clap-host/commit/f22ed540aa206c2452bdfb9011a76cf3b14cd9f2)
> This is how I think it could look like using the new `clap::helpers::Host` class. Maybe some details are still missing. What do you think @abique ?

That'd work for me. Thank you for doing this!

**Comment by abique** [comment]
> I think we could leave it like this for now except you want to have misbehavior checks from the beginning.

I can see that you still have some checkboxes opened. Whenever you think it is ready to be merged I'll give it a review and merge it.

**Comment by Trinitou** [comment]
Ok, @abique I looked through it and think it's ready enough for a review.

Cleaned up the checkboxes / comments. In the first comment I collected some ideas for future work / PRs.

**Comment by Trinitou** [comment]
One thing: In commit 5180427482efd8fc71ccb15e929b14c9ce498dc5 inside the `::from` method, I replaced the `std::cerr` + `std::terminate` with `throw` which I copied from the `clap-host` repo. Since the from method now is `noxcept`, this will cause a warning I think. Should we better switch back to std::terminate?

**Comment by abique** [comment]
@Trinitou Thank you for this work.
I'll review it in the afternoon.

File: include/clap/helpers/host.hh [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+      virtual bool implementsTimerSupport() const noexcept { return false; }
+      virtual bool timerSupportRegisterTimer(uint32_t periodMs, clap_id *timerId) noexcept { return false; }
+      virtual bool timerSupportUnregisterTimer(clap_id timerId) noexcept { return false; }
+
+      // clap_host_tail
+      virtual bool implementsTail() const noexcept { return false; }
+      static void tailChanged() noexcept {}
+
+      // clap_host_thread_check
+      virtual bool implementsThreadCheck() const noexcept { return false; }
```
Comment (abique) [inline include/clap/helpers/host.hh:81]: I'd make the thread check implementation mandatory maybe? :thinking:

Comment (abique) [reply include/clap/helpers/host.hh:81]: It isn't wrong to have it optional, but you loose 95% of the checks if you don't have the thread check. I think it isn't reasonable to start implementing a host without it.
Please make `threadCheckIsMainThread()` and `threadCheckIsAudioThread()` pure virtual and get rid of `implementsThreadCheck()`.

File: include/clap/helpers/host.hh [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+   protected:
+      Host(const char *name, const char *vendor, const char *url, const char *version);
+      virtual ~Host() = default;
+
+      /////////////////////////
+      // Methods to override //
+      /////////////////////////
+
+      // clap_host
+      virtual void requestRestart() noexcept {}
```
Comment (abique) [inline include/clap/helpers/host.hh:25]: Those 3 host methods aren't optional, so they should be pure virtual.

File: include/clap/helpers/host.hh [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+      virtual bool guiRequestHide() noexcept { return false; }
+      virtual void guiClosed(bool wasDestroyed) noexcept {}
+
+      // clap_host_latency
+      virtual bool implementsLatency() const noexcept { return false; }
+      static void latencyChanged() noexcept {}
+
+      // clap_host_log
+      virtual bool implementsLog() const noexcept { return false; }
+      virtual void logLog(clap_log_severity severity, const char *message) const noexcept {}
```
Comment (abique) [inline include/clap/helpers/host.hh:48]: default implementation here should write to `std::clog` maybe?

Comment (abique) [reply include/clap/helpers/host.hh:48]: Maybe `clog` isn't the right output, but something should go on the console at least by default.

Comment (abique) [reply include/clap/helpers/host.hh:48]: Ah actually I'm not sure, given that `implementsLog()` returns false by default.
The current code is fine :ok_hand:

**Pull request merged**